pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,208 @@
+package machinedeletionhooks
+
+import (
+	"context"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"time"
+
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	machinev1beta1client "github.com/openshift/client-go/machine/clientset/versioned/typed/machine/v1beta1"
+	machinelistersv1beta1 "github.com/openshift/client-go/machine/listers/machine/v1beta1"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type machineDeletionHooksController struct {
+	etcdClient                            etcdcli.EtcdClient
+	machineClient                         machinev1beta1client.MachineInterface
+	masterMachineLister                   machinelistersv1beta1.MachineLister
+	configMapListerForKubeSystemNamespace corev1listers.ConfigMapNamespaceLister
+	// machineAPIChecker determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	machineAPIChecker     ceohelpers.MachineAPIChecker
+	masterMachineSelector labels.Selector
+}
+
+// NewMachineDeletionHooksController reconciles machine hooks for master machines
+//
+// note:
+//  The Machine Deletion Hook this controller reconciles is a mechanism within the Machine API that allow this operator to hold up removal of a machine
+//  until a replacement member has been promoted to a voting member.
+func NewMachineDeletionHooksController(
```

Comment:
I am going to wire this controller in https://github.com/openshift/cluster-etcd-operator/pull/779

Reply:
> I am going to wire this controller in #779

Without wiring, we cannot verify that this code works.